File: libs/statemanager/src/main/kotlin/mono/state/controller/MouseInteractionController.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.controller

import mono.actionmanager.ActionManager
import mono.actionmanager.OneTimeActionType
import mono.actionmanager.RetainableActionType
import mono.common.post
import mono.graphics.geo.MousePointer
import mono.shape.selection.SelectedShapeManager
import mono.state.command.CommandEnvironment
import mono.state.command.MouseCommandFactory
import mono.state.command.mouse.HoverShapeManager
import mono.state.command.mouse.MouseCommand

/**
 * A controller class specific for mouse interaction.
 */
internal class MouseInteractionController(
    private val environment: CommandEnvironment,
    private val actionManager: ActionManager,
    private val requestRedraw: () -> Unit
) {
    val currentRetainableActionType: RetainableActionType
        get() = actionManager.retainableActionLiveData.value

    var currentMouseCommand: MouseCommand? = null
        private set

    private val lineConnectHoverShapeManager = HoverShapeManager.forLineConnectHover()
    private val hoverShapeManager = HoverShapeManager.forHoverShape()

    fun reset() {
        lineConnectHoverShapeManager.resetCache()
        hoverShapeManager.resetCache()
    }

    fun onMouseEvent(mousePointer: MousePointer) {
        if (mousePointer is MousePointer.Down || mousePointer is MousePointer.Up) {
            reset()
        }
        if (mousePointer is MousePointer.DoubleClick) {
            val targetedShape =
                environment.getSelectedShapes()
                    .firstOrNull { it.contains(mousePointer.boardCoordinate) }
            actionManager.setOneTimeAction(OneTimeActionType.EditSelectedShape(targetedShape))
            return
        }

        if (currentMouseCommand == null) {
            detectHoverShape(mousePointer)
        }

        val mouseCommand =
            MouseCommandFactory.getCommand(environment, mousePointer, currentRetainableActionType)
                ?: currentMouseCommand
                ?: return

        currentMouseCommand = mouseCommand

        environment.enterEditingMode()
        val commandResultType = mouseCommand.execute(environment, mousePointer)

        if (commandResultType == MouseCommand.CommandResultType.DONE) {
            environment.exitEditingMode(true)
        }

        if (commandResultType == MouseCommand.CommandResultType.DONE ||
            commandResultType == MouseCommand.CommandResultType.WORKING_PHASE2
        ) {
            lineConnectHoverShapeManager.resetCache()
            currentMouseCommand = null
            requestRedraw()
            // Avoid click when adding shape cause shape selection command
            post { actionManager.setRetainableAction(RetainableActionType.IDLE) }
        }
    }

    private fun detectHoverShape(mousePointer: MousePointer) {
        val (hoverShape, type) = when (currentRetainableActionType) {
            RetainableActionType.ADD_LINE -> {
                val hoverShape = lineConnectHoverShapeManager.getHoverShape(
                    environment,
                    mousePointer.boardCoordinate
                )
                hoverShape to SelectedShapeManager.ShapeFocusType.LINE_CONNECTING
            }

            RetainableActionType.IDLE -> {
                val hoverShape = hoverShapeManager.getHoverShape(
                    environment,
                    mousePointer.boardCoordinate
                )
                hoverShape to SelectedShapeManager.ShapeFocusType.SELECT_MODE_HOVER
            }

            RetainableActionType.ADD_RECTANGLE,
            RetainableActionType.ADD_TEXT -> {
                null
            }
        } ?: return
        environment.setFocusingShape(hoverShape, type)
        requestRedraw()
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/controller/mouse-interaction-controller.ts
import type { ActionManager } from "$mono/action-manager/action-manager";
import { OneTimeAction } from "$mono/action-manager/one-time-actions";
import { RetainableActionType } from "$mono/action-manager/retainable-actions";
import { ShapeFocusType } from "$mono/shape/selected-shape-manager";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import { MouseCommandFactory } from "$mono/state-manager/command/mouse-command-factory";
import { HoverShapeManager } from "$mono/state-manager/command/mouse/hover-shape-manager";
import { MouseCommand } from "$mono/state-manager/command/mouse/mouse-command";
import { MousePointer, MousePointerType } from "$mono/workspace/mouse/mouse-pointer";

/**
 * A controller class specific for mouse interaction.
 */
export class MouseInteractionController {
    private currentMouseCommand: MouseCommand | null = null;
    private lineConnectHoverShapeManager = HoverShapeManager.forLineConnectHover();
    private hoverShapeManager = HoverShapeManager.forHoverShape();

    constructor(
        private environment: CommandEnvironment,
        private actionManager: ActionManager,
        private requestRedraw: () => void,
    ) {
    }

    get currentRetainableActionType(): RetainableActionType {
        return this.actionManager.retainableActionFlow.value!;
    }

    reset(): void {
        this.lineConnectHoverShapeManager.resetCache();
        this.hoverShapeManager.resetCache();
    }

    onMouseEvent(mousePointer: MousePointer): void {
        if (mousePointer.type === MousePointerType.DOWN || mousePointer.type === MousePointerType.UP) {
            this.reset();
        }
        if (mousePointer.type === MousePointerType.DOUBLE_CLICK) {
            const targetedShape = this.findTargetedShape(mousePointer);
            this.actionManager.setOneTimeAction(OneTimeAction.EditSelectedShape(targetedShape));
            return;
        }

        if (this.currentMouseCommand === null) {
            this.detectHoverShape(mousePointer);
        }

        const mouseCommand =
            MouseCommandFactory.getCommand(this.environment, mousePointer, this.currentRetainableActionType)
            || this.currentMouseCommand;

        if (!mouseCommand) {
            return;
        }

        this.currentMouseCommand = mouseCommand;

        this.environment.enterEditingMode();
        const commandResultType = mouseCommand.execute(this.environment, mousePointer);

        if (commandResultType === MouseCommand.CommandResultType.DONE) {
            this.environment.exitEditingMode(true);
        }

        if (commandResultType === MouseCommand.CommandResultType.DONE ||
            commandResultType === MouseCommand.CommandResultType.WORKING_PHASE2) {
            this.lineConnectHoverShapeManager.resetCache();
            this.currentMouseCommand = null;
            this.requestRedraw();
            setTimeout(() => this.actionManager.setRetainableAction(RetainableActionType.IDLE), 0);
        }
    }

    private detectHoverShape(mousePointer: MousePointer): void {
        const [hoverShape, type] = (() => {
            switch (this.currentRetainableActionType) {
                case RetainableActionType.ADD_LINE: {
                    const hoverShape = this.lineConnectHoverShapeManager.getHoverShape(this.environment, mousePointer.boardCoordinate);
                    return [hoverShape, ShapeFocusType.LINE_CONNECTING];
                }
                case RetainableActionType.IDLE: {
                    const hoverShape = !this.isOnInteractionPoint(mousePointer)
                        ? this.hoverShapeManager.getHoverShape(this.environment, mousePointer.boardCoordinate)
                        : null;
                    return [hoverShape, ShapeFocusType.SELECT_MODE_HOVER];
                }
                case RetainableActionType.ADD_RECTANGLE:
                case RetainableActionType.ADD_TEXT:
                    return [null, null];
            }
        })();

        if (!hoverShape) {
            return;
        }

        this.environment.setFocusingShape(hoverShape, type);
        this.requestRedraw();
    }

    private findTargetedShape(mousePointer: MousePointer): AbstractShape | null {
        const selectedShapes = this.environment.getSelectedShapes();
        return Array.from(selectedShapes).find(shape => shape.contains(mousePointer.boardCoordinate)) ?? null;
    }

    private isOnInteractionPoint(mousePointer: MousePointer): boolean {
        switch (mousePointer.type) {
            case MousePointerType.DOWN: {
                const pointPx = mousePointer.clientCoordinate;
                return this.environment.getInteractionPoint(pointPx) !== null;
            }
            case MousePointerType.MOVE: {
                const pointPx = mousePointer.clientCoordinate;
                return this.environment.getInteractionPoint(pointPx) !== null;
            }
            default:
                return false;
        }
    }
}
